core/algo/threaded_loop.h

```diff
@@ -315,114 +316,123 @@ template <class OuterLoopType> struct ThreadedLoopRunOuter {
       return;
     }
 
-    std::mutex mutex;
     ProgressBar::SwitchToMultiThreaded progress_functions;
 
     struct Shared {
+      Shared(const Shared &) = delete;
+      Shared(Shared &&) = delete;
+      Shared &operator=(const Shared &) = delete;
+      Shared &operator=(Shared &&) = delete;
+      ~Shared() = default;
       Iterator &iterator;
       decltype(outer_loop(iterator)) loop;
-      std::mutex &mutex;
       FORCE_INLINE bool next(Iterator &pos) {
-        std::lock_guard<std::mutex> lock(mutex);
         if (loop) {
           assign_pos_of(iterator, loop.axes).to(pos);
           ++loop;
           return true;
         } else
           return false;
       }
-    } shared = {iterator, outer_loop(iterator), mutex};
-
-    struct PerThread {
-      Shared &shared;
-      typename std::remove_reference<Functor>::type func;
-      void execute() {
-        Iterator pos = shared.iterator;
-        while (shared.next(pos))
-          func(pos);
+      Iterator get_iterator() const {
+
+        MutexProtected<Shared> shared = {iterator, outer_loop(iterator)};
+
+        struct PerThread {
+          PerThread(const PerThread &) = default;
+          PerThread(PerThread &&) noexcept = default;
+          PerThread &operator=(const PerThread &) = delete;
+          PerThread &operator=(PerThread &&) = delete;
+          ~PerThread() = default;
+          MutexProtected<Shared> &shared;
+          typename std::remove_reference<Functor>::type func;
+          void execute() {
+            auto pos = shared.lock()->iterator;
+            while (shared.lock()->next(pos))
+              func(pos);
+          }
+        } loop_thread = {shared, functor};
+
+        auto threads = Thread::run(Thread::multi(loop_thread), "loop threads");
+
+        __manage_progress(&shared.lock()->loop, &threads);
```

Comment:
This part I'm not sure about. It seems that we are looking to a shared resource (the `loop`) outside the scope of this file to the `ProgressBar`. What guarantees that this is safe? @jdtournier

Reply:
Ok, so after discussing this should be safe, but we should revisit the logic of `ProgressBar` in future PRs.